File: extension/core/src/main/java/io/holunda/camunda/bpm/data/adapter/basic/ReadAdapterLockedExternalTask.java
package io.holunda.camunda.bpm.data.adapter.basic;

import io.holunda.camunda.bpm.data.adapter.ReadAdapter;
import io.holunda.camunda.bpm.data.adapter.VariableNotFoundException;
import org.camunda.bpm.engine.externaltask.LockedExternalTask;

import java.util.Optional;

/**
 * Read-write adapter for external task.
 *
 * @param <T> type of value.
 */
@SuppressWarnings("java:S1192")
public class ReadAdapterLockedExternalTask<T> implements ReadAdapter<T> {

  private final ReadAdapter<T> readAdapter;

  public ReadAdapterLockedExternalTask(LockedExternalTask lockedExternalTask, String variableName, Class<T> clazz) {
    readAdapter = new ReadWriteAdapterVariableMap<>(lockedExternalTask.getVariables(), variableName, clazz);
  }

  @Override
  public T get() {
    return readAdapter.get();
  }

  @Override
  @SuppressWarnings("unchecked")
  public Optional<T> getOptional() {
    return readAdapter.getOptional();
  }

  @Override
  public T getLocal() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public Optional<T> getLocalOptional() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public T getOrDefault(T defaultValue) {
    return readAdapter.getOrDefault(defaultValue);
  }

  @Override
  public T getLocalOrDefault(T defaultValue) {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

  @Override
  public T getOrNull() {
    return readAdapter.getOrNull();
  }

  @Override
  public T getLocalOrNull() {
    throw new UnsupportedOperationException("Can't get a local variable on an external task");
  }

}

File: extension/core/src/main/kotlin/io/holunda/camunda/bpm/data/adapter/basic/ReadAdapterLockedExternalTask.kt
package io.holunda.camunda.bpm.data.adapter.basic

import org.camunda.bpm.engine.externaltask.LockedExternalTask
import org.camunda.bpm.engine.variable.Variables
import java.util.*

/**
 * Read-write adapter for external task.
 *
 * @param [T] type of value.
 */
class ReadAdapterLockedExternalTask<T : Any?>(
  private val lockedExternalTask: LockedExternalTask,
  variableName: String,
  clazz: Class<T>
) : AbstractBasicReadWriteAdapter<T>(variableName, clazz) {
  private val value: Any?
    get() = Optional.ofNullable(lockedExternalTask.variables)
      .orElse(Variables.createVariables())[variableName]

  override fun getOptional(): Optional<T> {
    return Optional.ofNullable(
      getOrNull(
        value
      )
    ) as Optional<T>
  }

  override fun set(value: T, isTransient: Boolean) {
    throw UnsupportedOperationException("Can't set a variable on an external task")
  }

  override fun setLocal(value: T, isTransient: Boolean) {
    throw UnsupportedOperationException("Can't set a local variable on an external task")
  }

  override fun getLocal(): T {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOptional(): Optional<T> {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOrDefault(defaultValue: T): T {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun getLocalOrNull(): T {
    throw UnsupportedOperationException("Can't get a local variable on an external task")
  }

  override fun remove() {
    throw UnsupportedOperationException("Can't remove a variable on an external task")
  }

  override fun removeLocal() {
    throw UnsupportedOperationException("Can't remove a local variable on an external task")
  }
}
